_blog/deploying-laravel-to-kubernetes/index.md

```diff
@@ -0,0 +1,354 @@
+while learning and testing new setups but also indispensable when testing a 
+production scenario without having to lease any external services on the 
+cloud until you're ready to do so, fully knowing what will be needed and when.
+
+__Contexts__
+
+So, we use Kubectl to manage our local Minikube cluster and maybe a few 
+more clusters on the cloud. These different orchestration targets are contexts 
+which we can tell Kubectl to switch between.
+
+
+
+## Deploying the Laravel Application to Minikube
+
+The previous section, explaining some of the terms used when working with 
+Kubernetes may intimidate most of us, but let's not be. These are 
+simply words to compartmentalise different layers and attributes within a 
+deployment scenario.
+ 
+I have kept this example quite simple by removing state, i.e. no persistence 
+and thus no databases. In this example, we will use Minikube to locally deploy
+ a basic Laravel application across three nodes for load balancing.
+ 
+_Want to learn how to manage storage volumes and databases? We will be 
+writing a tutorial about this soon. Subscribe and be notified as
+ soon as new articles are published._
+ 
+__The application__
+
+I have provided a simple Laravel application which can be cloned from <a href="https://github.com/learnk8s/laravel-kubernetes-demo" target="_blank" rel="noopener" class="link">GitHub</a>. This is nothing more than a new 
+Laravel installation. Therefore you can follow this example using either the 
+demo application or simply create your own new Laravel application. To use 
+the demo application simply clone in your project's directory.
+
+```bash
+cd /to/your/working/directory
+git clone git@github.com:learnk8s/laravel-kubernetes-demo.git .
+```
+
+__Prerequisites__
+
+In order to follow with this demonstration, you will need the following 
+installed on your local system:
+
+1) <a href="https://docs.docker.com/install/" target="_blank" 
+rel="noopener">Docker</a>
+
+2) <a href="https://kubernetes.io/docs/tasks/tools/install-kubectl/" target="_blank" 
+rel="noopener">Kubectl</a>
+
+3) <a href="https://github.com/kubernetes/minikube/releases" target="_blank" 
+   rel="noopener">Minikube</a>
+
+_Having problems installing and running these applications on Windows? We will
+ be writing a tutorial on how to accomplish this very soon. Subscribe and be 
+notified as soon as new articles are published._
+
+__Docker image__
+
+Since this tutorial will be run locally, we can simply build a local Docker 
+Image from the `DockerFile` included in the example code. This `DockerFile` 
+is fairly basic:
+
+```bash
+FROM php:7
+RUN apt-get update -y && apt-get install -y openssl zip unzip git
+RUN curl -sS https://getcomposer.org/installer | php -- --install-dir=/usr/local/bin --filename=composer
+RUN docker-php-ext-install pdo mbstring
+WORKDIR /app
+COPY . /app
+RUN composer install
+
+CMD php artisan key:generate
+CMD php artisan serve --host=0.0.0.0 --port=8181
```

Comment:
AFAIK, you can only have 1 CMD. I'm not familiar with Laravel, what is this `php artisan key:generate` used for?

Reply:
These are simple CLI commands. The first one `php artisan key:generates` generates a random key/string to be used by the application's hashing algorithm.

This command exits once the key is generated and stored in an .env file.

`php artisan serve` loads the PHP's built-in server for serving the application.

Artisan is a cli tool provided by Laravel and it provides several tools such as code files' generation, testing helpers etc..

Reply:
If you generate the key as part of your CMD, every time the container starts a new key is generated. If you change your `CMD` for a `RUN` the command runs once at build time. I'm not sure when this should be run, so I'm just trying to explain what above.

Reply:
I didn't know that. You are perfectly right. The key should only be generated once per environment. I'll change it to `RUN`.